hathor/p2p/netfilter/matches.py

```diff
@@ -0,0 +1,177 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC
+from ipaddress import ip_network
+from typing import TYPE_CHECKING, Any, Set
+
+from structlog import get_logger
+from twisted.internet.defer import Deferred
+from twisted.internet.task import LoopingCall
+
+if TYPE_CHECKING:
+    from hathor.manager import HathorManager
+    from hathor.p2p.netfilter.context import NetfilterContext
+
+logger = get_logger()
+
+
+class NetfilterMatch(ABC):
+    """Abstract match class."""
+    def match(self, context: 'NetfilterContext') -> bool:
+        raise NotImplementedError
+
+
+class NetfilterMatchAll(NetfilterMatch):
+    """Always match."""
+    def match(self, context: 'NetfilterContext') -> bool:
+        return True
+
+
+class NetfilterMatchAnd(NetfilterMatch):
+    """Logic AND operation for two matches. When the first fails, the second is not executed."""
+    def __init__(self, a: NetfilterMatch, b: NetfilterMatch):
+        self.a = a
+        self.b = b
+
+    def match(self, context: 'NetfilterContext') -> bool:
+        if not self.a.match(context):
+            return False
+        if not self.b.match(context):
+            return False
+        return True
+
+
+class NetfilterMatchOr(NetfilterMatch):
+    """Logic OR operation for two matches. When the first matches, the second is not executed."""
+    def __init__(self, a: NetfilterMatch, b: NetfilterMatch):
+        self.a = a
+        self.b = b
+
+    def match(self, context: 'NetfilterContext') -> bool:
+        if self.a.match(context):
+            return True
+        if self.b.match(context):
+            return True
+        return False
+
+
+class NetfilterMatchIPAddress(NetfilterMatch):
+    """Match an IP address or network."""
+    def __init__(self, host: str):
+        """`host` can be either an ip address or a network.
+
+        Examples:
+            127.0.0.1
+            192.168.0.8
+            192.168.0.0/24
+            0.0.0.0/0 (match all addresses)
+        """
+        self.host = ip_network(host)
+
+    def match(self, context: 'NetfilterContext') -> bool:
+        if context.addr is None:
+            return False
+
+        if not hasattr(context.addr, 'host'):
+            return False
+
+        try:
+            host = ip_network(context.addr.host)
+        except ValueError:
+            return False
+
+        if self.host.version != host.version:
+            return False
+
+        if not self.host.supernet_of(host):
+            return False
```

Comment:
Apparently this is Python 3.7+ only. I don't think we should necessarily accelerate dropping support of Python 3.6 just for this, but at the very least we could leave this commented for a simpler way do this once we do drop 3.6

Reply:
Nice catch! Thanks!

I put an if-else block that run this code for 3.7+ and another code for 3.6. Can you review it agian?

Reply:
Even though `overlaps` and `supernet_of` have different semantics, they work the same for us because one host is always `/32`.

I just implemented your suggestion. I changed to `host = ip_address(...)` and used `host not in self.host`.